### DISCUSSION TITLE
How to inject functions to Templater (dynamically) from other plugins?

### DISCUSSION THREAD

**Discussion by flaper87** [SilentVoid13/Templater#970] [Help]
Hi,

I'm working on a plugin and I would like to have it depend on templater for several of the rendering parts. In order to do so, I was planning to add functions to templater when my plugin is invoked so that folks can use things like: 

`<% tp.plugin.something_something() %>`

Any suggestions on how to apprach/do this the best way?

**Comment by AB1908** [comment]
Can you explain what you mean by "depending on templater for rendering"? That is not a good use of Templater, if I understand correctly.

**Reply by flaper87** [reply]
Sorry, poor wording on my side. What I mean is that I would like people to be able to add functions from my plugin into their templater Templates. I would like to register these functions when the plugin is loaded rather than copying files (or asking users to copy files) into their custom JS directories.

I'm currently doing something like this (I will publish this plugin later this week): 

```typescript
    let section = '# 🤓 Stats';
    let parts = template_content.split(/# 🤓 Stats(?:\n|\r\n|\r)/);
    let final_template = template_content + section + "\n\n" + table + "\n\n";

    let outDir =
        this.app.vault.getAbstractFileByPath(this.settings.roastPathObsidian);
    await templater.templater.create_new_note_from_template(final_template,
                                                            outDir, filename);
```

I would like to leverage Templater capabilities instead of my plugin doing this search, split, etc, I know there are other things I could do to achieve this but it's a bit of a wasted considering how many Templater users are out there and how powerful it is. Just want to keep it DRY!

**Reply by AB1908** [reply]
Hmm, I'd probably ask on discord in the plugin-dev channel if I were you. I take it you want to expose APIs from your plugin. I think that's a fairly simple thing. However, the people writing templater scripts need to be aware that you're exposing APIs.

The other approach is that you depend on Templater to do the heavy lifting instead. This can be tricky. This is what I think you should ask about in the channel.
